SoftLink WritingToolsUI because of dependency cycle.

megangardner opened the pull request:
#### 45ace7335dc86189e103cc22bb017810d00fc03b
<pre>
SoftLink WritingToolsUI because of dependency cycle.
<a href="https://bugs.webkit.org/show_bug.cgi?id=275645">https://bugs.webkit.org/show_bug.cgi?id=275645</a>
<a href="https://rdar.apple.com/130038242">rdar://130038242</a>

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

WritingToolsUI links things that cause a dependency cycle, so soft link instead.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/cocoa/WritingToolsUISoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/WritingToolsUISoftLink.mm: Added.
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager initWithWebViewImpl:]):
(-[WKTextAnimationManager addTextAnimationTypeForID:withData:]):
(-[WKTextAnimationManager textPreviewsForChunk:completion:]):
(-[WKTextAnimationManager textPreviewForRect:completion:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:

Canonical link: <a href="https://commits.webkit.org/280165@main">https://commits.webkit.org/280165@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d9550da4a1e32ed11c4adf0f68c5bf44c096203

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55892 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/35216 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/8360 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58888 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/6324 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58018 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42837 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6520 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/45003 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4362 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57921 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/33107 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/48186 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26139 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29891 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5514 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4467 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/51860 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5785 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60477 "Built successfully") | 
| | [⏳ 🛠 vision ](https://ews-build.webkit.org/#/builders/visionOS-1-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5909 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/52434 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [⏳ 🛠 vision-sim ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/48255 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/51935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-1-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8262 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31045 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32129 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/33211 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31877 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/PAL/pal/PlatformMac.cmake:
```diff
@@ -191,6 +192,7 @@ list(APPEND PAL_SOURCES
     cocoa/UsageTrackingSoftLink.mm
     cocoa/VisionKitCoreSoftLink.mm
     cocoa/WebPrivacySoftLink.mm
+    cocoa/WritingToolsUISoftLink.h
```
This should be the source file instead
```suggestion
    cocoa/WritingToolsUISoftLink.mm
```

aprotyas commented on Source/WebCore/PAL/pal/cocoa/WritingToolsUISoftLink.h:
```diff
@@ -0,0 +1,52 @@
+SOFT_LINK_CLASS_FOR_HEADER(PAL, _WTTextPreview)
+
+
+
```
Nit: there are some extra newlines here

aprotyas reviewed (approved):
This looks right to me! Though you probably want to wait for a softlink expert 😛

whsieh commented on Source/WebCore/PAL/pal/cocoa/WritingToolsUISoftLink.h:
```diff
@@ -0,0 +1,52 @@
+SOFT_LINK_CLASS_FOR_HEADER(PAL, WTWritingTools)
+SOFT_LINK_CLASS_FOR_HEADER(PAL, WTWritingToolsViewController)
+
+SOFT_LINK_CLASS_FOR_HEADER(PAL, _WTTextPreviewAsyncSource)
```
Hm... it looks like `_WTTextPreviewAsyncSource` is an ObjC protocol, not a class. We should probably just remove this.

megangardner replied:
Ahh, this got added in an attempt to fix some errors, and not removed when it didn't work.

whsieh commented on Source/WebCore/PAL/pal/cocoa/WritingToolsUISoftLink.h:
```diff
@@ -0,0 +1,52 @@
+SOFT_LINK_CLASS_FOR_HEADER(PAL, WTWritingToolsViewController)
+
+SOFT_LINK_CLASS_FOR_HEADER(PAL, _WTTextPreviewAsyncSource)
+SOFT_LINK_CLASS_FOR_HEADER(PAL, _WTTextEffect)
```
Same with `_WTTextEffect`, this is just a protocol.

rr-codes commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -164,6 +164,7 @@
 #import <pal/cocoa/RevealSoftLink.h>
 #import <pal/cocoa/VisionKitCoreSoftLink.h>
 #import <pal/cocoa/TranslationUIServicesSoftLink.h>
+#import <pal/cocoa/WritingToolsUISoftLink.h>
```
Is this imoirt needed in this file?

whsieh replied:
Yes (it's used in not-yet-upstreamed code in WebKitAdditions)

megangardner replied:
And it will be more apparent in the update, since you already upstreamed that part.

whsieh commented on Source/WebCore/PAL/pal/cocoa/WritingToolsUISoftLink.h:
```diff
@@ -0,0 +1,52 @@
+SOFT_LINK_CLASS_FOR_HEADER(PAL, _WTTextPreviewAsyncSource)
+SOFT_LINK_CLASS_FOR_HEADER(PAL, _WTTextEffect)
+SOFT_LINK_CLASS_FOR_HEADER(PAL, _WTTextEffectView)
+SOFT_LINK_CLASS_FOR_HEADER(PAL, _WTTextPreviewSource)
```
This one (`_WTTextPreviewSource`) is also unreferenced in our code, and is also a protocol.

megangardner committed:
SoftLink WritingToolsUI because of dependency cycle.
https://bugs.webkit.org/show_bug.cgi?id=275645
rdar://130038242

Reviewed by Wenson Hsieh and Abrar Rahman Protyasha.

WritingToolsUI links things that cause a dependency cycle, so soft link instead.

* Source/WebCore/PAL/PAL.xcodeproj/project.pbxproj:
* Source/WebCore/PAL/pal/PlatformMac.cmake:
* Source/WebCore/PAL/pal/cocoa/WritingToolsUISoftLink.h: Added.
* Source/WebCore/PAL/pal/cocoa/WritingToolsUISoftLink.mm: Added.
* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/UIProcess/mac/WKTextAnimationManager.mm:
(-[WKTextAnimationManager initWithWebViewImpl:]):
(-[WKTextAnimationManager addTextAnimationTypeForID:withData:]):
(-[WKTextAnimationManager textPreviewsForChunk:completion:]):
(-[WKTextAnimationManager textPreviewForRect:completion:]):
* Source/WebKit/UIProcess/mac/WebViewImpl.mm:

Canonical link: https://commits.webkit.org/280165@main
